[BE] feat:  사용자 재생기록 저장 및 조회 #487

JO-eusan opened the pull request:
<!-- PR 제목 예시
ex. [BE] 로그인 API 구현
ex. [AN] 페이지 구현 -->

## 📌 변경 내용 & 이유
<!-- ex. 
- 사용자 로그인 기능 구현
- 로그인 실패 시 에러 메시지 반환 처리 추가 -->
- * 사용자 별 재생 기록을 저장하기 위한 `playing_history` 테이블 생성

컬럼명 | 타입 | 설명
-- | -- | --
id (PK) | BIGINT (PK) | 고유 ID
member_id (FK) | BIGINT (FK) | 회원 ID
hearit_id (FK) | BIGINT | 히어릿 ID
last_play_time | INT | 마지막 재생 위치 (초 단위)
is_finished | BOOLEAN | 콘텐츠 완청 여부 (Default: false)
updated_at | DATETIME | 마지막 청취 시간 (자동 업데이트)

* Create, Update,Read API 구현
   * Create: 사용자(member_id)가 히어릿(hearid_id)를 처음 재생하면 마지막 재생 위치가 요청으로 들어온다.
   * Update: 사용자(member_id)가 히어릿(hearid_id)를 다시 재생하면 마지막 재생 위치가 요청으로 들어온다.
     -> PlayingHistory 데이터 생성
   * Read: 기존 모든 히어릿 조회의 응답에 last_play_time을 추가한다. (기존 히어릿 조회 API 변경)

## 📸 스크린샷 (선택)
JPA Insert 사용 시, 100 이상의 요청만 들어와도 많은 시간이 걸린다는 문제를 확인하여 다음과 같이 개선했습니다.

- JPA Insert 
<img width="445" height="131" alt="스크린샷 2025-09-07 20 13 01" src="https://github.com/user-attachments/assets/12c387a4-242a-4691-acd5-91fdd95330c6" />

- JDBC Bulk Insert
<img width="445" height="131" alt="스크린샷 2025-09-08 17 32 38" src="https://github.com/user-attachments/assets/7c786378-62e4-4936-bdb6-6e8db182af5c" />


## 📢 논의하고 싶은 내용
<!-- 이러이러해서 이부분을 이러이러하게 구현했는데 이방식이 괜찮은지? -->

## 🧩 관련 이슈
<!-- 이슈 태그: #123 -->
<!-- 이슈 닫기: close #123 -->
close #487


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **신규 기능**
  - 재생기록 저장 API 추가(POST /api/v1/playing-histories), PlayingHistoryService/도메인·버퍼·스케줄러 및 bulk upsert 도입.
  - 검색·카테고리·상세·북마크 응답에 사용자 컨텍스트 기반 lastPlayTime·북마크 정보 포함(서비스·컨트롤러 시그니처 변경).

- **문서**
  - API 문서에 lastPlayTime 필드 및 재생기록 성공/권한없음 시나리오 추가.

- **테스트**
  - 도메인·서비스·컨트롤러 테스트에 재생기록·UserContext·lastPlayTime 검증 추가.

- **데이터베이스/운영**
  - playing_history 테이블 추가 및 last_play_time을 BIGINT로 마이그레이션.

- **예외/제한**
  - 버퍼 과부하 에러코드 및 BufferRequestException 추가.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

JO-eusan committed:
[BE] feat: 재생기록 저장을 위한 엔티티 생성 #487

- flyway 용 db migration 파일 생성

JO-eusan committed:
[BE] feat: 재생기록 저장 API #487

JO-eusan committed:
[BE] feat: 히어릿 조회 시 마지막 재생 시간 필드 추가 #487

- 단일 히어릿 상세 조회 API
- 검색 히어릿 조회 API
- 북마크 히어릿 조회 API

JO-eusan committed:
[BE] refactor: 기존 POST로 재생 기록 저장을 PUT으로 변경 #487

- 새로운 히어릿 재생 시, CREATE
- 기존 히어릿 재생 시, UPDATE

JO-eusan committed:
[BE] refactor: 기존 재생 기록 jpa insert를 jdbc bulk insert로 교체 #487

- 5초마다 또는 500개 이상 모이면 DB에 flush

Byesol committed:
refactor: colorCode 오타 수정 #487

Byesol commented on backend/src/test/java/com/onair/hearit/app/presentation/BookmarkControllerTest.java:
```diff
@@ -70,6 +70,8 @@ void readBookmarkHearitsTest() {
+                                                        fieldWithPath("content[].lastPlayTime").description(
+                                                                "히어릿 마지막 재생 시간(초)"),
                                                         fieldWithPath("content[].categoryColor").description(
                                                                 "해당 히어릿 카테고리 ColoCode")
```
수정했어용!

Byesol commented on backend/src/test/java/com/onair/hearit/app/presentation/PlayingHistoryControllerTest.java:
```diff
@@ -0,0 +1,145 @@
+                .header("Authorization", "Bearer " + token)
+                .contentType("application/json")
+                .body(request)
+                .filter(document("playing-history",
```
동일 api에 대해 문서가 하나가 되도록 수정 완료!

Byesol committed:
refactor: api 문서 스니펫 id 충돌 해결 #487

Byesol commented on backend/src/main/resources/db/migration/V202509071351__playing_history.sql:
```diff
@@ -0,0 +1,10 @@
+    `member_id` BIGINT NOT NULL,
+    `hearit_id` BIGINT NOT NULL,
+    `last_play_time` INT NOT NULL,
+    `is_finished` BOOLEAN NOT NULL,
+    `updated_at` DATETIME NOT NULL,
```
jpa에 의해서 자동으로 updataed_at 은 업데이트되기에 ddl문에는 필요 없어욧!

JO-eusan committed:
[BE] refactor: 마지막 재생 기록 타입 변경 #487

- Integer -> Long

JO-eusan committed:
[BE] refactor: PlayingHistory 검증 Service 계층에서 확인 #487

- queue insert 시, 검증하던 로직을 Service로 위임하여 안정성 확보
- PlayingHistory lastPlayTime 음수 검증 추가

JO-eusan committed:
[BE] refactor: 큐 사이즈 제한 및 응답 상태코드 분기 처리 #487

- 큐의 사이즈를 10000으로 설정
- 큐 사이즈가 꽉찬 경우에 429 응답

JO-eusan committed:
[BE] refactor: 불필요한 메서드 삭제 #487

- 재생 기록 수정 테스트 추가

JO-eusan committed:
[BE] refactor: JDBC batchUpdate 시 현재 시간 한 번만 호출 #487

Byesol committed:
[BE] refactor: n+1 문제 해결을 위해 fetch join 및 Map 이용하게 변경  #487

Byesol committed:
[BE] refactor: 카테고리별 히어릿 조회 시 lastPlayTime 반환하게 구현 #487
- map 방식을 활용해 n+1 문제 해결

JO-eusan committed:
[BE] fix: 버퍼에 기록 저장 전에 존재 여부 판단 #487

Byesol committed:
[BE] refactor: playtime 의 n+1 해결 방식을 map 활용 방식에서 쿼리로 변경 #487

JO-eusan committed:
[BE] refactor: 북마크 히어릿 조회에서 프로젝션 방식으로 N+1 문제 해결 #487

Byesol committed:
[BE] refactor: HearitWithPlayTime 인터페이스명 변경 #487

Byesol committed:
[BE] Merge branch 'feat/#487-playing-history' of https://github.com/woowacourse-teams/2025-hEARit into feat/#487-playing-history

JO-eusan committed:
[BE] test: 북마크 히어릿 조회에서 lastPlayTime 반환 값 테스트 추가 #487

Byesol committed:
[BE] refactor: 사용하지 않는 repository 메서드 제거 #487

Byesol committed:
[BE] test: 재생기록과 함께 카테고리별 히어릿 조회 테스트 작성 #487

Byesol committed:
[BE] test: 재생기록과 함께 카테고리별 히어릿 조회 테스트 #487

Byesol committed:
[BE] Merge branch 'feat/#487-playing-history' of https://github.com/woowacourse-teams/2025-hEARit into feat/#487-playing-history

JO-eusan committed:
Merge remote-tracking branch 'origin/develop-be' into feat/#487-playing-history

# Conflicts:
#	backend/src/main/java/com/onair/hearit/app/dto/response/BookmarkHearitResponse.java
#	backend/src/test/java/com/onair/hearit/app/presentation/BookmarkControllerTest.java

JO-eusan committed:
[BE] fix: BookmarkHearitResponse 생성을 정적 팩터리 메서드 활용 #487

JO-eusan committed:
[BE] docs: 재생 기록 저장 및 조회 문서화 설명 변경 #487

gabean13 commented on backend/src/main/java/com/onair/hearit/app/presentation/PlayingHistoryController.java:
```diff
@@ -0,0 +1,32 @@
+        }
+        return ResponseEntity.status(HttpStatus.OK).build();
+    }
+}
```
PUT은 리소스를 명시해서 사용하는게 표준적인 방법입니다.

현재 방법은 /playing-histories에 요청을 보내서 데이터를 생성하거나 업데이트하는 방식인데 
클라이언트에서 playing-history의 명확한 리소스 위치를 알지 못하기 대문에 POST가 더 적절한 것 같습니다. 

그리고 굳이 created랑 ok를 나눌필요없이 그냥 201 반환해도 될 것 같아요 클라이언트 쪽에서 생성이 아니라 수정이라는 부분을 몰라도 되지않나요?!
[참고링크](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Methods/PUT)

JO-eusan replied:
PUT은 자원이 명확하게 명시되었을 때 사용하는게 맞는 것 같습니다!
그렇다면 생성 유무보다는 POST를 성공했는지 여부만을 전달하면 좋을 것 같은데 어떻게 생각하시나요?
우선 POST + 200 OK로 변경해두었습니다.

gabean13 commented on backend/src/main/java/com/onair/hearit/app/infrastructure/scheduler/PlayingHistoryBuffer.java:
```diff
@@ -0,0 +1,41 @@
+
+    public void add(PlayingHistory playingHistory) {
+        if (!queue.offer(playingHistory)) {
+            throw new BufferRequestException("큐가 가득 차서 요청을 처리할 수 없습니다.");
```
이 부분은 추후에 좀 더 완전한 해결이 필요할 것 같아서 TODO 작성해주시면 좋을 것 같습니다.

JO-eusan replied:
재시도 로직 추가에 대한 Notion에 할일 및 목록에 추가해두겠습니다.

gabean13 reviewed (changes requested):
고생하셨어용

JO-eusan committed:
[BE] docs: 재생 기록 저장 API document 수정 #487

JO-eusan committed:
[BE] refactor: 재생 기록 저장 API 요청 메서드를 PUT에서 POST로 변경 #487

JO-eusan committed:
[BE] docs: 재생 기록 저장 API 상태코드 문서화 변경 #487

Byesol committed:
[BE] docs: 히어릿 컨트롤러 테스트에 헤더 추가 #487

Byesol committed:
Merge remote-tracking branch 'origin/feat/#487-playing-history' into feat/#487-playing-history

Byesol committed:
[BE] Merge remote-tracking branch 'origin/feat/#487-playing-history' into feat/#487-playing-history

Byesol committed:
Merge remote-tracking branch 'origin/feat/#487-playing-history' into feat/#487-playing-history

JO-eusan committed:
[BE] fix: 비회원 사용자 테스트 토큰 제거 #487

gabean13 reviewed (approved):
좋습니두
